### DISCUSSION TITLE
v3.0.0

### DISCUSSION THREAD

**Discussion by znorman-harris** [interactive-data-language/vscode-idl#2] [Announcements]
Official release of the extension! See CHANGELOG and main README for more information.

<hr /><em>This discussion was created from the release <a href='https://github.com/interactive-data-language/vscode-idl/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by cpiker** [comment]
Hi

Do you know how I can provide documentation for DLM functions in such a way that show up in vscode?  Most of the software I create for IDL takes the form of binary DLMs bundling existing C code.

Thanks,

**Comment by znorman-harris** [comment]
That's a great question, and one that we have thought about, but we don't have anything implemented yet.

We actually have a couple of use cases for this: SAVE files and anything that comes from external sources like DLLs and DLMs.

Would you be up for a workflow like this?

- We will create a new file type for the extension that indicates a "type" file for IDL
- Write normal IDL syntax for routine definitions within this file (with arguments, keywords, and types for inputs/outputs). There would be no logic within the routines themselves, just the type signature and you would be able to use the automated documentation tools to create the starting point for you 
- Include with your DLLs and DLMs and anyone else who uses the extension (or your code) would have access to the types

**Comment by cpiker** [comment]
That would be quite okay with me.  Currently I include IDL style comments in the [glue code](https://github.com/das-developers/das2dlm/blob/216eede4efce2f89b09f97cf28a4debec9225ed4/src/das2c_results.c#L172).  Maybe these could be copied out by a build script.  Right now I put to docs on a [wiki](https://github.com/das-developers/das2dlm/wiki/das2c_results) which is okay I guess, but they're not integrated into dev tools where they're needed.

Thanks for thinking about this.  It would be gratifying see help for DLM functions (and maybe structures?) within VSCode.

**Comment by znorman-harris** [comment]
Thanks for the link to your code! The docs that we support in the extension follow the IDLDoc RST format (we have some backwards compatibility with the original IDLDoc format).

That means there would need to be some changes to your docs style.

For reference, here's some links showing how we documented the IDL code that is included as a part of the extension:

Procedure:

https://github.com/interactive-data-language/vscode-idl/blob/0fa8b8caabfbe87d79df67afa1d2088b2aa5fd50/idl/vscode/notebooks/idlnotebook/idlnotebook__define.pro#L103-L113

Functions:

https://github.com/interactive-data-language/vscode-idl/blob/0fa8b8caabfbe87d79df67afa1d2088b2aa5fd50/idl/vscode/notebooks/idlnotebook/idlnotebook__define.pro#L204-L214

Structure definitions in the "__define" routine:

https://github.com/interactive-data-language/vscode-idl/blob/0fa8b8caabfbe87d79df67afa1d2088b2aa5fd50/idl/vscode/notebooks/idlnotebook/idlnotebook__define.pro#L444-L507

**Comment by cpiker** [comment]
Appreciate the pointers!  I don't mind switching over the comment style at all, just happy to have a supported standard to follow.  Will refer back to this discussion before I release the next version.

Cheers,

**Reply by znorman-harris** [reply]
@cpiker I know it has been a while, but we finally got to add in what we call Routine Definition files which allow you to have auto-complete and hover-help in VSCode. 

TLDR: Normal PRO code without any logic in routines. Docs format following IDL Doc. If you have format on save enabled, and auto-doc, then the hard work will be done for you.

You can take a look [here](https://interactive-data-language.github.io/vscode-idl/types/routine_definition_files.html) at our first pass.
